Update default registry to xpkg.upbound.io

jbw976 opened the pull request:
### Description of your changes

This PR updates the default value for the registry that Crossplane's package manager will install packages from. The default was previously `index.docker.io`, and the new default will be `xpkg.upbound.io`.  Details about the motivation for this change can be ready in #5012.

In addition to the simple change of the default value, it was necessary to update the package manager logic that creates deployment manifests that run providers and functions. That logic is now aware of the default registry so it can include it into manifests it creates when the user has not specified an explicit registry source. 

Feedback is welcome ASAP so we can gain confidence that these changes are thorough and we aren't missing any scenarios.

Fixes #5012 

### Testing

The following scenarios have been tested so far and their behavior is described briefly.  Links to gists with specific testing steps are provided for the more complicated scenarios in the table.

| Pkg Registry Set | Default Set | Upgrade Performed | Scenario | Behavior | 
| --------------------------- | ------------------| -------------------------------- | --------- | -------- |
| ❌  | ❌  | ❌  | Install new package w/o explicit registry specified: Providers, Functions, and Configurations | ✅ Package is installed from xpkg.upbound.io, can create resources OK |
| ✅  | ❌  | ❌  | Install new package with explicit registry specified  | ✅ Package is installed from specified registry, can create resources OK | 
| ❌  | ❌  | ✅  | Package installed on v1.14 w/o explicit registry specified, then Crossplane upgraded to this PR</br>Steps: [gist link](https://gist.github.com/jbw976/793210f26f2dfbb458a41d5dc437ce7e) | ⚠️  Behavior change: the package manager will attempt to update the existing package to the new default registry location. If the package does not exist on the new registry location, the package will become unhealthy. | 
| ✅  | ❌  | ✅  | Package installed on v1.14 w/ explicit registry specified, then Crossplane upgraded to this PR</br>Steps: [gist link](https://gist.github.com/jbw976/83695c4a31f7149bfbaea8249ed638c8) | ✅ Package stays healthy after the upgrade, existing resources stay healthy, can create new resources OK | 
| ❌  | ✅  | ✅  | Package installed on v1.14 w/o explicit registry specified, then Crossplane is upgraded to this PR with `--registry=index.docker.io` to keep previous behavior</br>[gist link](https://gist.github.com/jbw976/094cdfab0eb502e8369259f605761bda) | ✅ The behavior from previous releases of Crossplane remains intact, package stays healthy and installed from dockerhub. |
| ❌  | ❌  | ❌  | Dependencies in a package that are specified w/o an explicit registry [gist link](https://gist.github.com/jbw976/afd49f0a5fdc59a2c7aff559b00fb5ac) | ✅   Package and its dependency are installed from the default registry OK and everything reaches a healthy state. | 
| ✅  | ❌  | ❌  | `DeploymentRuntimeConfig` specifies an image w/ no registry set that overrides the image the package was installed with [gist link](https://gist.github.com/jbw976/f518c7c8f2d1d384c88f75f078d5119a)  | ✅  The image specified in the `DeploymentRuntimeConfig` is used exactly as is, no default is taken into account. If this image doesn't exist then the package will not get healthy, but we will honor exactly what the user specifies in the DRC. | 
| ❌  | ❌  | ❌  | `ControllerConfig` specifies an image w no registry set that overrides the image the package was installed with [gist link](https://gist.github.com/jbw976/fff6dc20c1042079182e55579fa5a1a2) | ✅  The image specified in the `ControllerConfig` is used exactly as is, no default is taken into account. Same behavior as `DeploymentRuntimeConfig`. | 


The most challenging behavior will be the likely common scenario of having existing packages installed implicitly from dockerhub (by previously not explicitly specifying a registry), then upgrading Crossplane to the functionality in this PR. Since the package manager will essentially try to upgrade the package from its previous location (dockerhub) to its new location on the default registry (xpkg.upbound.io), the package can become unhealthy if it doesn't exist there.

**Note** that this will happen for all the `crossplanecontrib` packages on dockerhub because that organization is called `crossplane-contrib` on `xpkg.upbound.io` (note the lack of a `-` on dockerhub), and therefore those packages won't exist.

We will absolutely need clear messaging about this change in the release notes and we should write a blog post too.  For anyone that has an environment where this behavior will cause problems, **ANY** of these options would be recommended:

* update your packages so they all have an explicit registry specified, e.g. `index.docker.io/myorg/mypkg`
* copy/push your packages to `xpkg.upbound.io` so they exist there in addition to dockerhub
* keep the previous behavior by using `--set args='{"--registry=index.docker.io"}'` during the helm upgrade of Crossplane v1.15

I have: <!--You MUST either [x] check or [ ] ~strike through~ every item.-->

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added or updated unit tests.
- [x] Added or updated e2e tests.
- [x] Linked a PR or a [docs tracking issue] to [document this change]: https://github.com/crossplane/docs/issues/680
- [ ] ~Added `backport release-x.y` labels to auto-backport this PR.~

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/master/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/master/contributing#checklist-cheat-sheet

jbw976 commented on internal/controller/pkg/revision/runtime_test.go:
```diff
@@ -125,30 +124,6 @@ func TestRuntimeManifestBuilderDeployment(t *testing.T) {
 				})),
 			},
 		},
-		"ProviderDeploymentWithImageOverride": {
```
the logic that this test was covering now actually lives at a higher level - this logic is now covered by `TestGetProviderImage`

jbw976 commented on internal/controller/pkg/revision/runtime_test.go:
```diff
@@ -353,30 +328,12 @@ func TestRuntimeManifestBuilderDeployment(t *testing.T) {
 				want: deploymentFunction(functionName, functionRevisionName, functionImage),
 			},
 		},
-		"FunctionDeploymentWithImageOverride": {
```
the logic that this test was covering now actually lives at a higher level - this logic is now covered by `TestGetFunctionImage`

jbw976 commented on internal/controller/apiextensions/controller/options.go:
```diff
@@ -35,10 +35,6 @@ type Options struct {
 
-	// Registry is the default registry to use when pulling containers for
-	// Composition Functions
-	Registry string
```
i believe this was leftover from the alpha implementation of functions that had specific machinery with a function runner and rootless containers. Now it more or less relies on the package manager to download/install/run a function in a deployment.

negz replied:
You're right. Looks like it was added in https://github.com/crossplane/crossplane/pull/4278.

jbw976 commented:
I've marked this PR as ready for review since I have made updates in both core and RBAC manager and tested a lot of scenarios. The results of all testing so far is included in the table now, along with lists to gists for details on some of the more complicated scenarios.

I will perform more testing, but this is ready for review and feedback 🙇‍♂️

jbw976 commented:
The scenario for having dependencies inside a package that don't explicitly specify a dependency seems to be broken - we should definitely fix this.

Details: https://gist.github.com/jbw976/afd49f0a5fdc59a2c7aff559b00fb5ac

Error snippet:
```
2024-01-19T20:14:21Z	DEBUG	crossplane	cannot fetch dependency package tags	{"controller": "packages/lock.pkg.crossplane.io", "request": {"name":"lock"}, "uid": "1e6ef810-09cb-4be4-97fd-ad348e7c3a79", "version": "825", "name": "lock", "error": "GET https://index.docker.io/v2/crossplane-contrib/provider-helm/tags/list?n=1000: UNAUTHORIZED: authentication required; [map[Action:pull Class: Name:crossplane-contrib/provider-helm Type:repository]]"}
```

jbw976 commented:
A commit has been pushed to fix the scenario where dependencies are declared in a package without specifying the registry.  The testing table has been updated ✅

negz commented on cmd/crossplane/rbac/rbac.go:
```diff
@@ -53,7 +53,7 @@ var KongVars = kong.Vars{
 		},
 		", "),
-	"default_registry": name.DefaultRegistry,
+	"rbac_default_registry": xpkg.DefaultRegistry,
```
Did this need to be renamed, or were you just going for something more explicit? (Just curious.)

jbw976 replied:
actually this change was needed for the RBAC manager to be able to have its own default too. Before this change, 
The RBAC command was using the exactly same variable name (conflict) as the Core command was. So when we tell the kong cli package to parse all the vars, core is the last one listed and I think that makes it win the conflict, forcing RBAC manager to use Core's default.

This variable name change avoids this conflict and lets RBAC have its own default, as was originally intended I believe.

negz commented on internal/controller/pkg/revision/runtime_function.go:
```diff
@@ -192,3 +199,19 @@ func functionServiceOverrides() []ServiceOverride {
+// getFunctionImage determines a complete function image, taking into account a
+// default registry. If the function meta specifies an image, we have a
+// preference for that image over what is specified in the package revision.
+func getFunctionImage(functionMeta *pkgmetav1beta1.Function, pr v1.PackageRevisionWithRuntime, defaultRegistry string) (string, error) {
```
```suggestion
func getFunctionImage(fm *pkgmetav1beta1.Function, pr v1.PackageRevisionWithRuntime, defaultRegistry string) (string, error) {
```

This might be a good opportunity to switch to a shorter variable name. `functionMeta` seems redundant given it's mostly repeating information that already exists in the type. AFAICT it has limited scope in the sense discussed at https://github.com/crossplane/crossplane/tree/master/contributing#use-descriptive-variable-names-sparingly.

jbw976 replied:
Good call! I'll make the variable name shorter here and in a couple other similar places.

negz commented on internal/controller/pkg/revision/runtime_provider.go:
```diff
@@ -192,7 +203,7 @@ func (h *ProviderHooks) Deactivate(ctx context.Context, pr v1.PackageRevisionWit
 }
 
-func providerDeploymentOverrides(providerMeta *pkgmetav1.Provider, pr v1.PackageRevisionWithRuntime) []DeploymentOverride {
+func providerDeploymentOverrides(providerMeta *pkgmetav1.Provider, pr v1.PackageRevisionWithRuntime, image string) []DeploymentOverride {
```
Same comment here WRT a shorter variable name.

negz reviewed (commented):
This looks good to me, thanks @jbw976! If possible I'd like @turkenh to approve before we merge though, since he's the most up-to-date with the package manager code.

I searched for uses of `ParseReference` that this PR doesn't touch. It mostly looks good:

* `cmd/crank/xpkg/build.go` has a comment explaining why overriding the default registry wouldn't make sense there.
* Various other places already plumb down the default correctly.
* `CopyImageToRegistry` in the E2E tests is unused. It can probably be deleted (see https://github.com/crossplane/crossplane/pull/4323).

The only one I'm not sure about is `internal/initializer/installer.go`, that has a bunch of places that explicitly pass `name.WithDefaultRegistry("")`. Any idea what's up with that?

turkenh reviewed (approved):
The PR looks good to me! The PR description and gist files was super helpful 🙌 
I have an observation regarding the dependency resolution that I detailed below. I don't think it should be a blocker for this PR and we could handle it separately. I'll leave the decision to you 🙇‍♂️ 

---
Should we take default registry into account during package resolution as well? I reused your custom function with provider helm dependency to test what happens when we have the dependent package already installed but without defaulting and I believe there could be some inconsistency that is worth handling. 

I installed the below manifests:

First this one:

```yaml
apiVersion: pkg.crossplane.io/v1
kind: Provider
metadata:
  name: provider-helm
spec:
  package: xpkg.upbound.io/crossplane-contrib/provider-helm:v0.16.0
```

Once the above is ready, this one: 

```yaml
apiVersion: pkg.crossplane.io/v1beta1
kind: Function
metadata:
  name: jbw976-jfn
spec:
  package: jbw976/jfn:v0.1.10
```

where function with package `jbw976/jfn:v0.1.10` has a dependency to provider with package `crossplane-contrib/provider-helm`. Given that we are defaulting to `xpkg.upbound.io`, the dependency is already satisfied. However, the package manager dependency resolution didn't take this into account and tried to deploy provider-helm again ending up with the state below:

<img width="1712" alt="Screenshot 2024-01-20 at 15 03 33" src="https://github.com/crossplane/crossplane/assets/9900707/db66bdb7-60b9-4f91-aaa1-7f218f2994b4">

Looks like registry defaulting wasn't taken into account during dependency resolution before, but could be ok with the old default, i.e. `index.docker.io` considering it is the default everywhere. So, it might reveal some already existing problem now.

jbw976 commented on internal/controller/pkg/revision/runtime_provider.go:
```diff
@@ -220,10 +231,6 @@ func providerDeploymentOverrides(providerMeta *pkgmetav1.Provider, pr v1.Package
-	if providerMeta.Spec.Controller.Image != nil {
-		image = *providerMeta.Spec.Controller.Image
-	}
 	do = append(do, DeploymentRuntimeWithOptionalImage(image))
```
Good point @turkenh, I just tried that scenario out and you are totally correct. I'll update the testing table with this scenario so it's clear there, and I'll also add a comment to the code here (and also in runtime_function.go) so it's clear for folks in the code base in the future.

jbw976 commented:
> The only one I'm not sure about is internal/initializer/installer.go, that has a bunch of places that explicitly pass name.WithDefaultRegistry(""). Any idea what's up with that?

@negz the best I can tell is that function is pre-installing a set of providers/configurations that can (apparently) be supplied to the init cmd/container: 

* https://github.com/crossplane/crossplane/blob/release-1.14/cmd/crossplane/core/init.go#L37-L38 
* https://github.com/crossplane/crossplane/blob/release-1.14/internal/initializer/installer.go#L55

I imagine that it is purposefully installing with no default registry so that EXACTLY what the user specifies for these packages is what gets installed? We're taking a similar approach for packages specified in `DeploymentRuntimeConfig`.  I thiiiiiiink this is all good if I'm reading it right 😅

jbw976 commented on internal/controller/pkg/revision/runtime_function_test.go:
```diff
@@ -570,3 +583,80 @@ func TestFunctionDeactivateHook(t *testing.T) {
+		reason string
+		args   args
+		want   want
+	}{
```
cool! added another test case for both functions and providers that makes sure the registry specified on the package is honored, even if its different than the default registry.

jbw976 commented:
> I have an observation regarding the dependency resolution that I detailed below. I don't think it should be a blocker for this PR and we could handle it separately

> Looks like registry defaulting wasn't taken into account during dependency resolution before, but could be ok with the old default, i.e. index.docker.io considering it is the default everywhere. So, it might reveal some already existing problem now.

@turkenh great observation, thank you for your thoroughness and insight here man!! 🙇‍♂️ 🙇‍♂️ 🙇‍♂️ 

After thinking through this a bit more, I agree that it wasn't taken in account before either so it's not a strict regression but may become more common with the default change.  I think this scenario is possible but not overwhelmingly likely, so let's go ahead and fix this in an upcoming PR after we land this functionality here.

I've opened #5270 to track this and I've included it in the v1.15 milestone.

jbw976 commented:
i think it's time to abandon trying to get a `provider-nop` release out of its `release-0.1` branch - the publish-artifacts step even fails too:  https://github.com/crossplane-contrib/provider-nop/actions/runs/7608427037/job/20717566360. I think that branch has bit rotted too much to deal with now 😇 

It may be wiser/smoother to try to get a new build out of the `release-0.2` branch that @negz already modernized, then have the [provider upgrade e2e test](https://github.com/crossplane/crossplane/blob/master/test/e2e/pkg_test.go#L92) run an upgrade of provider-nop from `v0.2.0` --> `v0.2.1`.  that still meets the intent of the test IMO. I'll try that in the morning 😂

jbw976 commented:
@negz @phisco @turkenh I think this is good to go now! All the e2e tests are passing except for realtime-compositions `TestPropagateFieldsRemovalToXR/TestPropagateFieldsRemovalToXR/CreateClaim`, which I believe to be flaky: #5228
https://github.com/crossplane/crossplane/actions/runs/7617613918/job/20747009792?pr=5261

Let me know if anything else is needed for merging here! 🙇‍♂️

negz commented on internal/controller/pkg/revision/runtime_function.go:
```diff
@@ -116,7 +120,14 @@ func (h *FunctionHooks) Post(ctx context.Context, pkg runtime.Object, pr v1.Pack
 	sa := build.ServiceAccount()
-	d := build.Deployment(sa.Name, functionDeploymentOverrides(functionMeta, pr)...)
+
+	// determine the function's image, taking into account the default registry
```
Nit: I'd prefer if comments were written with proper sentence grammar, i.e. caps and periods. We haven't been super consistent on this but I believe it's the predominant style in the codebase.

negz commented on internal/controller/pkg/revision/runtime_provider.go:
```diff
@@ -136,7 +140,14 @@ func (h *ProviderHooks) Post(ctx context.Context, pkg runtime.Object, pr v1.Pack
 	sa := build.ServiceAccount()
-	d := build.Deployment(sa.Name, providerDeploymentOverrides(providerMeta, pr)...)
+
+	// determine the providers's image, taking into account the default registry
```
Same nit. 😄

negz reviewed (approved):
Looks good to me, thank you @jbw976!

jbw976 commented:
@negz your last feedback has been incorporated, I've squashed all the feedback commits into a single tidy commit, and I've rebased off the latest in master.

If CI is green then let's merge!! 🙏
